src/util/retry.rs

```diff
@@ -0,0 +1,255 @@
+/*
+ copyright: (c) 2013-2020 by Blockstack PBC, a public benefit corporation.
+
+ This file is part of Blockstack.
+
+ Blockstack is free software. You may redistribute or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License or
+ (at your option) any later version.
+
+ Blockstack is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY, including without the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::io::prelude::*;
+use std::io;
+use std::io::{Read, Write};
+
+use util::log;
+use util::hash::to_hex;
+
+/// Wrap a Read so that we store a copy of what was read.
+/// Used for re-trying reads when we don't know what to expect from the stream.
+pub struct RetryReader<'a, R: Read> {
+    fd: &'a mut R,
+    buf: Vec<u8>,
+    do_buffer: bool,
+    i: usize
+}
+
+impl<'a, R: Read> RetryReader<'a, R> {
+    pub fn new(fd: &'a mut R) -> RetryReader<'a, R> {
+        RetryReader {
+            fd: fd,
+            buf: vec![],
+            do_buffer: true,
+            i: 0
+        }
+    }
+
+    pub fn set_position(&mut self, offset: usize) -> () {
+        if offset <= self.buf.len() {
+            self.i = offset
+        }
+        else {
+            self.i = self.buf.len()
+        }
+    }
+
+    pub fn position(&self) -> usize {
+        self.i
+    }
+
+    pub fn disable_bufferring(&mut self) -> () {
+        self.do_buffer = false;
+    }
+
+    fn read_and_buffer(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let nr = self.fd.read(buf)?;
+        if self.do_buffer {
+            self.buf.extend_from_slice(buf);
+            self.i += buf.len();
+        }
+        Ok(nr)
+    }
+}
+
+impl<'a, R: Read> Read for RetryReader<'a, R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```

Comment:
The interactions between `disable_bufferring`, `i`, and `buf` are problematic in this struct. I think in its existing usage, it won't matter (though the panicking behavior could be bad now in certain http cases), but it could be problematic down the line:

```rust
    #[test]
    fn test_bad_retry_reader() {
        use std::io::Read;
        let buf = vec![1,2,3,4,5,6,7,8,9,10,11,12];
        let mut cursor = io::Cursor::new(&buf);
        let mut rrd = RetryReader::new(&mut cursor);

        let mut a = [0,0,0,0];
        rrd.read(&mut a);
        println!("A: {:#?}", a);
        // a => 1,2,3,4
        // i = 4, buf= [1,2,3,4];
        rrd.disable_bufferring();
        rrd.read(&mut a);
        println!("A: {:#?}", a);
        // a => [5,6,7,8];
        // i = 4, buf = [1,2,3,4];
        rrd.set_position(0);
        let mut b = [0,0,0,0,0];
        rrd.read(&mut b[0..4]);
        rrd.read(&mut b[4..]);
        println!("B: {:#?}", b);
        // b => [1,2,3,4,9];
        // i = 4, buf = [1,2,3,4];
        assert!(false);
    }
```

There's two problems: first, the buff "b" there is receiving malformed stream data -- it gets 1,2,3,4,9. second, if I had passed "b" in as a length 5 buff, the read call would have panicked, because copy_from_slice panics when the buf lengths don't match exactly.

Seeing as how `disable_bufferring` is only ever used by the auth serialization code, it might be easier just to get rid of that buffer on/off functionality, and switch the auth serialization code to use the chain method of readers, e.g.:

in auth.rs:
```rust
    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<TransactionSpendingCondition, net_error> {
        // peek the hash mode byte
        let hash_mode_u8 : u8 = read_next(fd)?;
        let chain_reader = [hash_mode_u8];
        let mut chained = chain_reader.chain(fd);

        let cond = { 
            if SinglesigHashMode::from_u8(hash_mode_u8).is_some() {
                let cond = SinglesigSpendingCondition::consensus_deserialize(&mut chained)?;
...
```

As for the panicking `copy_from_slice`, &[u8] implements read, which handles that logic:

```
...
            if self.i < self.buf.len() {
                // consume from inner buffer
                let bytes_copied = (&self.buf[self.i..]).read(buf)?;
                self.i += bytes_copied;
                bytes_copied
            }
...
```

Reply:
Thanks, didn't know about `.chain()`.  Will use that instead of `RetryReader`, since `RetryReader` was created to make it possible to peek bytes from a stream.

Reply:
Addressed in 951ee88aa and 675296cb5